### ISSUE TITLE
[HOLD for payment 2022-03-22] [$500] Connection drop is not detected - Reported by: @Santhosh-Sellavel 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#6059]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Steps to simulate connect loss:
2. Turn on mobile internet on phone.
3. Turn on hotspot/internet sharing on your phone.
4. Connect your laptop/pc to the network shared from mobile.
5. Use the new Expensify app.
6. Now disable mobile internet in the phone, now its connection drop but our app doesn’t detect it.
7. You can verify opening any user chat, we can't see the offline message warning.

## Expected Result:
Connection drop should be detected on all circumstances. 

## Actual Result:
Connection drop is not detected and the app should online status. 

## Workaround:
None, going back online fixes the issue. 

## Platform:

Where is this issue occurring?

- Web
- Desktop App
 
**Version Number:** 1.1.9-1

**Reproducible in staging?:** Yes 
**Reproducible in production?:** Yes 

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![image](https://user-images.githubusercontent.com/44479856/138804276-1fc30a45-69b9-4b30-8c06-14043cbd7890.png)

**Expensify/Expensify Issue URL:**

**Issue reported by:** @Santhosh-Sellavel 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1634766280030400

**Job:** https://www.upwork.com/jobs/~0129cfc8254ac008fa

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
@kevinksullivan Eep! 4 days overdue now. Issues have feelings too...

**Comment by kevinksullivan** [comment]
I was unable to reproduce this on web. The app responded  correctly when I turned off hotspot.

![image](https://user-images.githubusercontent.com/22508290/139787695-e208ec04-1f77-4e01-8224-087e692733fb.png)

**Comment by Santhosh-Sellavel** [comment]
Sorry couldn’t able to get here early @kevinksullivan
1. Create hotspot using mobile internet
2. Connect the Laptop to the hotspot
3. Now Turn off the `mobile internet` `not the created hotspot`!

**Comment by MelvinBot** [comment]
@kevinksullivan Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by kevinksullivan** [comment]
When you say `mobile internet` do you mean wifi , cellular, or something else?

**Comment by Santhosh-Sellavel** [comment]
Cellular data @kevinksullivan

**Comment by kevinksullivan** [comment]
Ok, I did that and could not reproduce. Feel free to post a video of this if I am missing something, but I am going to close until then.

**Comment by Santhosh-Sellavel** [comment]
@isagoico Can you help me here. Can you reproduce this now. Can you add attachments if reproducible.

**Comment by Santhosh-Sellavel** [comment]
@kevinksullivan 

I think maybe you used iPhone for hotspot sharing. When you turn off the cellular data internet hotspot also turns off on iPhone. I verified using iPhone not able to reproduce using iPhone as a hotspot device.

Try with android you will be able to reproduce! Attached a video [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1637339317339300?thread_ts=1634766280.030400&cid=C01GTK53T8Q) 

In the video, the device is connected to an android device sharing the internet hotspot. Once the internet is turned down in the android device, the iPhone wifi setting detects no internet. But our app fails to detect.

**Comment by isagoico** [comment]
I was able to reproduce the issue with the same steps as before in build 1.1.14-2 (Desktop)

**Comment by kevinksullivan** [comment]
https://www.upwork.com/jobs/~0129cfc8254ac008fa

**Comment by MelvinBot** [comment]
Triggered auto assignment to @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @pecanoro (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by kidroca** [comment]
This is related: https://github.com/Expensify/App/issues/5337

Long story short both issues are because we rely on `navigator.onLine` which is unreliable. 
 `navigator.onLine` being `true` means only that you are connected to some network - weird or wireless, and not that the network has internet connection 

Here I propose to double check when our network status changes: https://github.com/Expensify/App/issues/5337#issuecomment-943543012
This might require a dedicated endpoint to serve for internet checks or we can use an existing one:
- when a request fails without a response - we set the app as offline and send another request to double check 
- when the app connection status changes to online or offline - we send a request to double check 

Depending on the result is the double check we might need to set the network status again 

If we're not using a dedicated endpoint we can just make a request for something simple like user current time zone for the double check

**Comment by mallenexpensify** [comment]
@kidroca , do you think this issue and https://github.com/Expensify/App/issues/5337 and should be _combined_ or separate issues?

**Comment by kidroca** [comment]
Yes, the main problem is connection state is not tracked reliably (enough) in both issues

**Comment by pecanoro** [comment]
Ok, so I am closing this one then since it's basically the same issue.

**Comment by mallenexpensify** [comment]
Reopening this in favor of trying to fix this issue which is more common than the similar issue here #5337 .  Reasoning here
https://github.com/Expensify/App/issues/5337#issuecomment-1002762626

**Comment by mallenexpensify** [comment]
Created a new Upwork job, doubled price to $500
https://www.upwork.com/jobs/~01cc84a0bf57e59bb5

**Comment by K4tsuki** [comment]
# proposal

Actually App have reliable internet connection detection in `PusherConnectionManager`.
We could inject `NetworkConnection.setOfflineStatus();`

in `Pusher.registerSocketEventCallback`:

https://github.com/Expensify/App/blob/9f3dc928acb6bddf72c851ee8332cc628f86010a/src/libs/PusherConnectionManager.js#L24-L39

When `disconnected` or `error` event came we set `setOfflineStatus` to true. when `connected` event came we set `setOfflineStatus` to false.

Here we have reliable `connected to internet` event.

**Comment by kidroca** [comment]
> Here we have more reliable `connected to internet` event, compared to currently used `NetInfo` using state change event, even with checking on `isInternetReachable` status.

How come one is more reliable than the other? 
`isInternetReachable` performs an internet request while Pusher is only connecting to the Pusher service.
Having the app in the background would disconnect from pusher, making you "offline", then when focused again, there would be some time to re-connect to pusher and be online "again" (e,g, auth token refresh -> reconnect to pusher)

We can use certain events like Pusher disconnecting or a network request timing out to force `NetInfo` re-check the status, but ultimately let `NteInfo` decide. Otherwise we're having multiple points that can change online state

**Comment by K4tsuki** [comment]
https://github.com/pusher/pusher-js#react-native

 > pusher-js depends on NetInfo.

So pusher is using NetInfo under the hood, but their implementation on using this NetInfo is more reliable and more correct than current our implementation on using this NetInfo (As you can see there is many bugs around offline and online state of the Apps).
Their library is used by many packages and by many users, it is more mature.  


> `isInternetReachable` performs an internet request while Pusher is only connecting to the Pusher service.

Have you seen any log about this request. Where is the target request to check on `isInternetReachable`? Pusher is not just some random server. We use Pusher in our App, so pooling on Pusher server is not random internet check. Our app is rely on this server to function properly. If for some reason Pusher server is down (edge case), it is might be better to set the state  of the  App offline. Because this might be confusing the user (or maybe showing message that server is down)

> Having the app in the background would disconnect from pusher, making you "offline", then when focused again, there would be some time to re-connect to pusher and be online "again" (e,g, auth token refresh -> reconnect to pusher)

For faster check we don't need to wait until it pusher state is `connected`, but until `connecting` this is fast and we could check online by http request to expensify server, and set the offline status accordingly.

> We can use certain events like Pusher disconnecting or a network request timing out to force NetInfo re-check the status, but ultimately let NteInfo decide. Otherwise we're having multiple points that can change online state

I think using Pusher event is enough here. I have tested it turning off and on wifi, mobile data and connect to no internet wifi, Pusher responds is good. But current App implementation on NetInfo usage doesn't respond correctly.

------------------------------------------
If you are not convinced enough, let the reporter (Santhosh-Sellavel) or someone to test this concept on this problem and similar problems on offline and online state of the App.

**Comment by kidroca** [comment]
`isInternetReachable` doesn't have to make a request on the platforms that have a native way of knowing whether they have internet access 
But if it does have to make request it's [configured here](https://github.com/react-native-netinfo/react-native-netinfo#netinfoconfiguration): 

> Property | Type | Default | Description
> -- | -- | -- | --
> reachabilityUrl | string | https://clients3.google.com/generate_204 | The URL to call to test if the internet is reachable. Only used on platforms which do not supply internet reachability natively.

If the Pusher team can make Net Info work for them it only proves we're not using it the best way
A solution that works doesn't automatically make it the best. 
If a simple solution like switching to `isInternetReachable` works I would prefer it

Is it enough that Pusher is disconnected to consider we're offline and block any of our requests - we only use it to receive messages, for everything else we use the http api. 

Instead of blocking all network requests because Pusher disconnected we can
- force a re-check 
- or let http request through and only set ourselves as offline if they fail

Making a re-check instead instead of instantly assuming we're offline seems the most simple and covering a wider range of cases

**Comment by kidroca** [comment]
This might sound a bit radical alternative but 

Why just simply never be offline 😅 
Or rather, indicate to users they might be offline (like we do), but don't block any requests because of it - when requests fail they can be put back to the queue for a retry

**Comment by kevinksullivan** [comment]
Still in planning, not overdue.

**Comment by K4tsuki** [comment]
> If the Pusher team can make Net Info work for them it only proves we're not using it the best way
A solution that works doesn't automatically make it the best.
If a simple solution like switching to isInternetReachable works I would prefer it

No it is  not simple solution. I think there is something wrong with our usage on NetInfo. It is  not just adding check on `isInternetReachable`. Something  is wrong with current implementation. I think the more simpler solution is to depend on currently available Pusher implementation and double check (like you said) using http request to expensify server.

> Is it enough that Pusher is disconnected to consider we're offline and block any of our requests?

Pusher socket update connection state if there is change in network or connectivity state. The state change is something like this:

From Connected to disconnected:
1. From connected to disconnected to internet, state change from `connected` to `connecting`
2. After some trying to connect fail, state change from `connecting` to `unavailable`

For disconnected to connected:
1. state change from `unavailable` to `connecting`
2. If internet reachable and connected to pusher server, state change from `connecting` to `connected`
3. If pusher is  not able to `connect`, state change from `connecting` back to `unavailable`


>  we only use it to receive messages, for everything else (including sending messages) we use the http api.

Sending IOU actions also depends on pusher too, because server create the report action and then we receive the report action via Pusher. I am still not yet checks on other functions that depends on pusher. [Please remember this is maybe the edge case when pusher server is down]


> Instead of blocking all network requests because Pusher disconnected we can
>    force a re-check
>    or let http request through and only set ourselves as offline if they fail
> Making a re-check instead of instantly assuming we're offline seems the most simple and covering a wider range of cases

This can be done to double check that we are really not connected to internet.

> This might sound a bit radical alternative but
> Why just simply never be offline sweat_smile
Or rather, indicate to users they might be offline (like we do), but don't block any requests because of it - when requests fail they can be put back to the queue for a retry, eventually succeeding

For queued request, this might be fine (the log will show many http request error though).
But App need clear state of the offline status, I think we depend on this state for displaying something on some components.

----------------------------------------------------------------

So there are three way:
1. Using pusher connection state and double check when connection is `connecting` state.
2. Fixing current implementation on NetInfo usage (pusher source code might be good reference)
3.  `never be offline` alternative ( Or rather, indicate to users they might be offline (like we do), but don't block any requests because of it - when requests fail they can be put back to the queue for a retry, eventually succeeding). I am not sure about this.

So which one is best here? If you want to go to second route, I will investigate the pusher source code.

**Comment by kidroca** [comment]
> No it is not simple solution. I think there is something wrong with our usage on NetInfo

If there's something wrong with our side we should fix it, or do you mean we should stop using `NetInfo` altogether and replace existing logic with Pusher connected/disconnected events?

Our entire usage of `NetInfo` is this

https://github.com/Expensify/App/blob/7af06bcc0a05edc3f38febb7967814a60604a56b/src/libs/NetworkConnection.js#L59-L62

### Proposal:

1. Switch that to `isInternetReachable` and see if this fixes the problem
2. Add a `NetInfo.fetch` to recheck state, when `AppState` changes to `active`  
    - [connection state events won't be triggered in the background](https://github.com/react-native-netinfo/react-native-netinfo#switching-between-different-wi-fi-does-not-send-events-in-ios)
3. Expose a method like `NetworkConnection.recheck` that's getting triggered from places like Pusher or Fetch errors 

`NetworkConnection.recheck` can force a `NetInfo.fetch` (optionally we can use our own api as the internet reachability endpoint)
Both `NetInfo.fetch` and `NetInfo.addEventListener` would use the same callback:
```js
function receiveNetworkState(state) {
  Log.info(`[NetworkConnection] NetInfo isConnected: ${state && state.isInternetReachable}`);

  if (state && _.isBoolean(state.isInternetReachable)) {
    setOfflineStatus(!state.isInternetReachable);
  }  
}
```

**Comment by parasharrajat** [comment]
Thanks, @K4tsuki and @kidroca for putting up a great discussion. I have a few questions for both of you but I am not a subject matter expert for this. 

@K4tsuki, I agree with Peter that the pusher can get disconnected on the backend and how will we continue to sync the background data? 

> So pusher is using NetInfo under the hood, but their implementation on using this NetInfo is more reliable and more correct than current our implementation on using this NetInfo (As you can see there is many bugs around offline and online state of the Apps).
Their library is used by many packages and by many users, it is more mature.

How can you say that their implementation is more reliable? Have you inspected the code? Can you show us Why?

> For faster check we don't need to wait until it pusher state is connected, but until connecting this is fast and we could check online by http request to expensify server, and set the offline status accordingly.

Here you are saying that we check on Pusher and then also do HTTP polling to determine the status. How will you determine that polling is needed to check the validity of internet status?

Out of all your [three approaches](https://github.com/Expensify/App/issues/6059#issuecomment-1007016068), which one do you think is best and why?

I got your plan but I don't see a firm decision of what are you suggesting to do. Could you please lay down the full plan divided into steps with technical changes you will do to achieve this?

---

@kidroca Waiting on HTTP poll requests could create a significant delay especially when the internet is not reliable. This could have various side effects. 

Based on the info https://github.com/pusher/pusher-js#react-native. 
I see that they use Netinfo and as per @K4tsuki they can consistently check the internet status. Also, the doc says that 
> The fallbacks available for this runtime are HTTP streaming and polling.

So it seems to me that they have everything in place. But I have seen this discussed before that Pusher is slow so we moved to NetInfo. 

I see a plan but I would prefer that we determine the internet status from native APIs for the platforms that expose network reachability APIs rather than depending on HTTP polling and the third party and fall back to HTTP polling when required for the platforms where we don't have dedicated APIs.

**Comment by kidroca** [comment]
> @kidroca Waiting on HTTP poll requests could create a significant delay especially when the internet is not reliable. This could have various side effects.

Something I said might have thrown you off I never suggested any kind of polling

`NetInfo` is not constantly running `internetReachability` - only when it has to. And if it can - it will use native capabilities to determine whether internet is reachable

> I see a plan but I would prefer that we determine the internet status from native APIs for the platforms that expose network reachability APIs rather than depending on HTTP polling

Yes - stick with `NetInfo` it does exactly that - `isInternetReachable` uses native APIs and only performs a reachability request when such APIs are not available

So in the end `NetInfo` would do HTTP polling if it has no other option
- when internet was reachable it will do a request every 60 seconds to check whether it's still reachable (this is [configurable](https://github.com/react-native-netinfo/react-native-netinfo#netinfoconfiguration))
- when internet becomes unreachable it will do a request every 5 seconds to check whether we're back online (this is also [configurable](https://github.com/react-native-netinfo/react-native-netinfo#netinfoconfiguration)) 

These seems like sensible defaults to me. I don't see anything more we could do on platforms that don't have a choice beside triggering a recheck manually if we have insights we might be offline (pusher disconnect / fetch failing)

**Comment by parasharrajat** [comment]
Oh, great. I see what you mean now. To me, this seems a good approach to what @kidroca has proposed. 

Depending on the Pusher socket connection maybe not be the best way to handle this especially when it also usages NetInfo. There could be more layers around the Pusher socket. Another reason is we should not rely on third-party servers when we have native solutions. 

@kidroca Have you done some tests with the reproduction steps? 

I will wait to see what  @K4tsuki has to say about it. Also, for this 
> How can you say that their implementation is more reliable? Have you inspected the code? Can you show us Why?

**Comment by kidroca** [comment]
> @kidroca Have you done some tests with the reproduction steps?

I did some time ago before the ticket was closed, and manage to recreate. But I haven't tested whether `isInternetReachable` would fix the problem, I didn't know it existed back then 
This reminded me that for the Desktop App, we're actually not using `NetInfo` but a simple stub that just using `navigator.onLine`

https://github.com/Expensify/App/blob/431b60d09f4e41c7532707454625dbe2c3c22c21/src/libs/NetInfo/index.desktop.js#L14-L16

So there's no internet reachability detection at all on that platform.
Meaning that connecting to a tethered connection but disable the internet and the app would not catch it thus: "Connection drop is not detected" (this ticket) 

https://github.com/Expensify/App/blob/431b60d09f4e41c7532707454625dbe2c3c22c21/src/libs/NetInfo/index.desktop.js#L2-L5

The ticket cited here was fixed in v7.1.3: https://github.com/react-native-netinfo/react-native-netinfo/issues/450 and we're already on v7.1.3 so it might be prime time to delete `NetInfo/index.desktop.js`
Or update the implementation to use the Electron API even though it doesn't sound very promising: https://www.electronjs.org/docs/latest/api/net#properties
> **net.online**
> A return value of false is a pretty strong indicator that the user won't be able to connect to remote sites. However, a return value of true is inconclusive...

**Comment by parasharrajat** [comment]
I agree, we should use the native electron package if we have one. They must be using OS-level APIs to determine that and thus I believe it should work.

> Or update the implementation to use the Electron API even though it doesn't sound very promising

we can quickly try and see if that works.

**Comment by K4tsuki** [comment]
Testing something, please wait. Thank you.

**Comment by K4tsuki** [comment]
I have done some testing...

On android:
1. The slow to connect on pusher-js is from the interval of offline polling. The interval is long, but  there is no configuration option to change this. Even if we set  the app status to online, then reconnect happens and all reports have been fetched again after reconnect, If pusher  start slower (15000 and increasing on default pusher js configuration), there is a chance that message will not displayed. Because Pusher is off and will not receive new event in that time. I explain this again in below section.
2. I have tested NetInfo in android. On my phone, turning on and off wifi doesn't trigger NetInfo state change event. Both in App and in Pusher. But pusher provide fallback or pooling mechanism on offline mode. So there is already polling mechanism in pusher on offline mode, but the interval is long and hence to `connect` in the worst case scenario pusher will need around 60 second. Set by this line:

https://github.com/pusher/pusher-js/blob/ac35c732233f52157cbda08276511ed29b89d378/src/runtimes/isomorphic/default_strategy.ts#L62-L66

I don't find the config to change this value. And make change to this value is necessary. I explain it in the section below. But summary is: "If pusher  start slower there is a chance that message will not displayed. Because Pusher is off and user will not get new event (message, etc) in that time. I explain this again in section below."

3. If you want to fix the Netinfo, App need to poll connection on offline and online mode. This just double the polling in the app that pusher already do.
4. [info] On android turning wifi off will immediately put pusher state in `connecting` mode, but in web pusher state keep on  `connected` until some time, but  this can be configured by `pongTimeout`  option when creating new Pusher instance.

---------------------------------------------
> How can you say that their implementation is more reliable? Have you inspected the code? Can you show us Why?

Because pusher provide fallback mechanism. Here on android the NetInfo state change doesn;t getting triggered, but pusher provides polling not only on online but on offline state too. 
This is what I have noticed that pusher success reconnecting on network change in my early proposal, but I don't know at that time that pusher provide offlline polling too. And also noticed that NetInfo state change doesn't get triggered.

> @K4tsuki, I agree with Peter that the pusher can get disconnected on the backend and how will we continue to sync the background data?

If pusher get disconnected, then App will not get new event from pusher. User will miss messages, etc. That is why App should be put to `offline` in this situation, to refetch reports and other data after App state back to `online` (here pusher server is up). If we keep forcing to put App in `online` state, there user might be miss message(s) and they will not know (because there is no reports refetch).

>  For faster check we don't need to wait until it pusher state is connected, but until connecting this is fast and we could check online by http request to expensify server, and set the offline status accordingly.
> Here you are saying that we check on Pusher and then also do HTTP polling to determine the status. How will you determine that polling is needed to check the validity of internet status?

I have made a mistake here. Pusher state after `unavailable` is `connected`, but it can be configured on `unavailableTimout`.
I made a mistake here, I thought `connecting` mean network state is change.
Http polling is not needed, because pusher already  provide  this. Please look at above section.

> Out of all your three approaches, which one do you think is best and why?
I thought Kidroca is proposal reviewer, so I just summarize solutions that he suggest (number 2 and 3). 
But my proposal is keep on using pusher (number 1).

> I got your plan but I don't see a firm decision of what are you suggesting to do. Could you please lay down the full plan divided into steps with technical changes you will do to achieve this?

Please wait. Summary is, to (-) listen pusher connection state change and set app offline state accordingly. 
(-) I will add some parameters when creating pusher instance
(-) I think the change in this [line](https://github.com/pusher/pusher-js/blob/ac35c732233f52157cbda08276511ed29b89d378/src/runtimes/isomorphic/default_strategy.ts#L62-L66) is really necessary to make pusher reconnect faster.

---------------------------------------

So to really tells whether the internet is connected or not, app need to keep polling on certain interval. Pusher has provide this in online and offline mode. But on  offline mode in the current configuration is slow. This can be changed by forking the repo. This is necessary and explanation is in above section.

**Comment by parasharrajat** [comment]
> I have tested NetInfo in android. On my phone, turning on and off wifi doesn't trigger NetInfo state change event. Both in App and in Pusher. But pusher provide fallback or pooling mechanism on offline mode. So there is already polling mechanism in pusher on offline mode, but the interval is long and hence to connect in the worst case scenario pusher will need around 60 second. Set by this line:

I was looking in the pusher code but I did not find any code that implements the fallback polling. Could you please help me find it? You are talking about it so maybe link us in the code, to better understand it. 

> If pusher get disconnected, then App will not get new event from pusher. User will miss messages, etc. That is why App should be put to offline in this situation, to refetch reports and other data after App state back to online (here pusher server is up). If we keep forcing to put App in online state, there user might be miss message(s) and they will not know (because there is no reports refetch).

When we receive notifications, we sync some data which can't rely on Pusher.

> This can be changed by forking the repo. This is necessary and explanation is in above section. 

I am not sure if we want to do that. Pusher is a very active codebase that relies on its servers. I am 90% sure we don't want to fork it and maintain ourselves. If you feel this is a bug in their code/ want to suggest a feature request on the lib, then you will have to submit an issue and PR to their repository. But I find it very unlikely that they will change their config for us. 

Looking at the information shared above, I feel we should do the following.

1. Use native internet detection as much as possible. It will require switching the Desktop implementation to use Election's exports etc.
2. I am leaning towards @kidroca  's proposal if that works. He is yet to submit test results. 


I will bring someone more experienced in this discussion cc: @marcaaron @tgolen

**Comment by K4tsuki** [comment]
> (because there is no reports refetch)
I have made a mistake here. Actually there is report refetch after Pusher subscribed  to user  event:

https://github.com/Expensify/App/blob/74cf52fea3a69f1c3503215defdd75de94b0881c/src/libs/actions/User.js#L294-L299

There is  `NetworkConnection.triggerReconnectionCallbacks`. 
Actually if pusher subscribe later, we will not miss message, because there is reconnection callback that refetch all reports after pusher get connected and subscribed to user event.

**Comment by tgolen** [comment]
Hi! Happy to stop by and chime in, but I'll admit, there is a lot to read through here, and I'm not sure what is necessary to read or not. Can someone help me by summing up where this is at and how I can help?

**Comment by parasharrajat** [comment]
There are two main proposals. 

Here are a few comments that contain most of them. They can explain the rest, I am not sure what to pick from all that.

@kidroca  https://github.com/Expensify/App/issues/6059#issuecomment-1007365923 https://github.com/Expensify/App/issues/6059#issuecomment-1007614463.

@K4tsuki https://github.com/Expensify/App/issues/6059#issuecomment-1007016068 https://github.com/Expensify/App/issues/6059#issuecomment-1008728484

**Comment by kidroca** [comment]
My proposal revolves on trying to make NetInfo work for us and make it the source of truth for our connectivity/internet state

@K4tsuki proposes to monitor Pusher events. But I'm not sure whether his proposal is to 
- totally get rid of NetInfo and rely on Pusher connected/disconnected state
- or have both NetInfo and Pusher dictate our connectivity

My point is if we should keep using NetInfo we should make it work for us
We can still force NetInfo to make a re-check if
- pusher disconnects 
- `fetch` fails with no server response (e.g. request timeout)

---

And we no longer need this file because the issue it mentions is already fixed (and we're on the fixed version): https://github.com/Expensify/App/blob/431b60d09f4e41c7532707454625dbe2c3c22c21/src/libs/NetInfo/index.desktop.js#L2-L5

**Comment by K4tsuki** [comment]
@tgolen Summary of my proposal (disregarding maybe an edge case where Pusher server going down):

I think to determine application `offline` or `online` status correctly, we need some kind of polling on internet connection.
Because we already using pusher that have polling mechanism on `offline` and `online` state, I utilizing pusher `connection` `state change` to determine application `offline` or `online` status.

Pros:
1. Don't add additional burden to `server` and `client` by polling connection 
2. Put state of application `offline` when pusher off (because some function on application depend on pusher, so it will not confuse user)

Cons:
1. Pusher slower to reconnect than simple `HTTP request` after `offline` state (On android, at least that is what I test) because offline interval of offline polling of pusher set by this line:
https://github.com/pusher/pusher-js/blob/ac35c732233f52157cbda08276511ed29b89d378/src/runtimes/isomorphic/default_strategy.ts#L62-L66
If we could change that line, we could reconnect faster. But maybe it is not convenient to forking the repo.
2. Maybe some others cons that I am not aware of, maybe @kidroca should point out other cons.

**Comment by tgolen** [comment]
I really like the idea of having a single source of truth for this mechanism, so I am leaning toward @kidroca's proposal. I know that Pusher isn't always the most reliable on mobile, especially with a flaky connection or backgrounding the app. Also, I would prefer if we don't rely on any polling mechanism.

**Comment by parasharrajat** [comment]
Thanks, Tim. I think it's good to have a slack discussion here before we proceed. 

Is anyone interested in summarizing both the approaches on Slack?

**Comment by kidroca** [comment]
I've just stumbled on this in the Network tab (web/desktop):
![image](https://user-images.githubusercontent.com/12156624/152840274-89316256-6e51-4c80-8285-155fb19ea519.png)

These calls originate from: `webpack:///./node_modules/@react-native-community/netinfo/lib/module/internal/internetReachability.js`
![image](https://user-images.githubusercontent.com/12156624/152841115-f7da4ba9-5ea6-41a8-b6e6-9ad551f9adfb.png)

Looks like they happen every minute, regardless of whether you're online or not - I thought polling would start only after there are clues that you might be offline ... 

This is staging. I guess we don't see these calls in DEV, because there's no point, net-info only enables them for a production build 

So it seems net-info does tests for internet connection, but we're not using the result - we're using `state.isConnected` which might be `true`, while `state.isInternetReachable` would be `false` resulting in "Connection drop not detected"

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @pecanoro, @kevinksullivan, @parasharrajat eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by pecanoro** [comment]
It seems Melvin went crazy yesterday 🤯 

Anyways, after re-reading everything, I think we should go with @kidroca's proposal since I think it would be a better approach, @parasharrajat do you agree with it?

**Comment by pecanoro** [comment]
Awesome! @kidroca, feel free to apply to the job in Upwork

**Comment by MelvinBot** [comment]
📣 @kidroca You have been assigned to this job by @pecanoro!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by kidroca** [comment]
Applied on Upwork, 
You can expect a PR submitted by 2022-02-18 EOD

**Comment by kidroca** [comment]
PR is ready: https://github.com/Expensify/App/pull/7825

**Comment by kevinksullivan** [comment]
PR still in review.

**Comment by PauloGasparSv** [comment]
Very similar issue https://github.com/Expensify/App/issues/7975 was fixed by the P.R. found in this issue.

**Comment by kevinksullivan** [comment]
@kidroca @Santhosh-Sellavel looks like the job expired so I created a new one and invited you. Once you submit I will hire and pay out the $500. Thanks!

https://www.upwork.com/jobs/~0115bb5b01faa7d4b8

**Comment by parasharrajat** [comment]
@kevinksullivan Should I apply on the same for C+?

**Comment by kidroca** [comment]
This is actually completed, I've applied to the new ticket and received payment

**Comment by pecanoro** [comment]
Closing then!

**Comment by parasharrajat** [comment]
But Reporter and C+ are still not paid @kevinksullivan.

**Comment by Santhosh-Sellavel** [comment]
@kevinksullivan
 
I too received payment but for this, only C+ is missed out here for @parasharrajat.

**Comment by parasharrajat** [comment]
@kevinksullivan Bump.

**Comment by kevinksullivan** [comment]
Apologies @parasharrajat . Just hired, please accept and I'll get this payment over.

**Comment by kbecciv** [comment]
@kidroca @kevinksullivan We reported the issue during Regression https://github.com/Expensify/App/issues/8690

**Comment by kidroca** [comment]
I've taken a look and it seems the Network code was (or is being) refactored 
At the time this issue closed the functionality worked correctly

If you want me to investigate further, please tag me on the new ticket
